threescale/client/http/client.go

```diff
@@ -0,0 +1,81 @@
+package http
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/3scale/3scale-go-client/threescale/client"
+)
+
+const (
+	authzEndpoint   = "/transactions/authorize.xml"
+	authRepEndpoint = "/transactions/authrep.xml"
+	reportEndpoint  = "/transactions.xml"
+)
+
+const (
+	defaultBackendUrl = "https://su1.3scale.net:443"
+	defaultTimeout    = 10 * time.Second
+
+	serviceIDKey = "service_id"
+
+	enableExtensions = "3scale-options"
+	// limitRemainingHeaderKey has a value set to the remaining calls in a current period
+	limitRemainingHeaderKey = "3scale-limit-remaining"
+	// limitResetHeaderKey has a value set to an integer stating the amount of seconds left for the current limiting period to elapse
+	limitResetHeaderKey = "3scale-limit-reset"
+
+	httpReqErrText = "error building http transaction"
+
+	// a parsable time format used to convert Ruby time to time type
+	timeLayout = "2006-01-02 15:04:05 -0700"
+)
+
+var (
+	errHttpReq = errors.New(httpReqErrText)
+)
+
+// NewClient returns a pointer to a Client providing some verification and sanity checking
+// of the backendURL input. backendURL should take one of the following formats:
+//	* http://example.com - provided scheme with no port
+//	* https://example.com:443 - provided scheme and defined port
+func NewClient(backendURL string, httpClient *http.Client) (*Client, error) {
+	url, err := verifyBackendUrl(backendURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		backendHost: url.Hostname(),
+		baseURL:     backendURL,
+		httpClient:  httpClient,
+	}, nil
+}
+
+func (c *Client) Authorize(req client.Request) (AuthorizeResponse, error) {
+	return nil, nil
+}
+
+func (c *Client) AuthRep(req client.Request) (AuthorizeResponse, error) {
+	return nil, nil
+}
+
+func (c *Client) Report(req client.Request) (ReportResponse, error) {
+	return nil, nil
+}
+
+// Verifies a custom backend is valid
+func verifyBackendUrl(urlToCheck string) (*url.URL, error) {
+	backendURL, err := url.ParseRequestURI(urlToCheck)
+	if err == nil {
+		scheme := backendURL.Scheme
+		if scheme != "" && scheme != "http" && scheme != "https" {
```

Comment:
if `ParseRequestURI` accepts whatever `scheme` was used, perhaps it can be used by the HTTP client? (ie. `http3`)